--- rsyslog-1.19.0/syslogd.c.orig	2007-08-21 16:45:10.000000000 +0200
+++ rsyslog-1.19.0/syslogd.c	2007-08-21 16:45:28.000000000 +0200
@@ -5650,11 +5650,11 @@ static void mainloop(void)
 #endif
 #endif
 
-	errno = 0;
-	FD_ZERO(&readfds);
-	maxfds = 0;
-
 	while(!bFinished){
+	        errno  = 0;
+	        maxfds = 0;
+	        FD_ZERO (&readfds);
+
 		/* first check if we have any internal messages queued and spit them out */
 		processImInternal();
 
