diff -up rsyslog-3.14.1/syslogd.c.legacyOpts rsyslog-3.14.1/syslogd.c
--- rsyslog-3.14.1/syslogd.c.legacyOpts	2008-04-07 17:15:45.000000000 +0200
+++ rsyslog-3.14.1/syslogd.c	2008-04-07 17:15:49.000000000 +0200
@@ -1740,6 +1740,7 @@ void legacyOptsHook(void)
 
 	while(pThis != NULL) {
 		if(pThis->line != NULL) {
+			errno = 0;
 			errmsg.LogError(NO_ERRCODE, "Warning: backward compatibility layer added to following "
 				        "directive to rsyslog.conf: %s", pThis->line);
 			conf.cfsysline(pThis->line);
@@ -3111,7 +3112,7 @@ int realMain(int argc, char **argv)
 					legacyOptsEnq((uchar *) "ModLoad imuxsock");
 					bImUxSockLoaded = 1;
 				}
-				legacyOptsEnq((uchar *) "OmitLocaLogging");
+				legacyOptsEnq((uchar *) "OmitLocalLogging");
 			} else {
 				fprintf(stderr, "error -o is no longer supported, use module imuxsock instead");
 			}
