diff -up rsyslog-3.12.3/modules.c.objinfo rsyslog-3.12.3/modules.c
--- rsyslog-3.12.3/modules.c.objinfo	2008-03-18 18:21:48.000000000 +0100
+++ rsyslog-3.12.3/modules.c	2008-03-18 18:22:14.000000000 +0100
@@ -700,7 +700,6 @@ Release(char *srcFile, modInfo_t **ppThi
 		/* we have a zero refcount, so we must unload the module */
 		dbgprintf("module '%s' has zero reference count, unloading...\n", pThis->pszName);
 		modUnlinkAndDestroy(&pThis);
-		*ppThis = NULL; /* nobody can access it any longer! */
 	}
 
 	RETiRet;
