diff -up rsyslog-3.21.3/action.c.clock rsyslog-3.21.3/action.c
--- rsyslog-3.21.3/action.c.clock	2008-08-28 10:23:10.000000000 +0200
+++ rsyslog-3.21.3/action.c	2008-08-28 10:37:24.000000000 +0200
@@ -577,6 +577,10 @@ dbgprintf("NTH: conf: %d, actual %d\n", 
 
 	dbgprintf("Called action, logging to %s\n", module.GetStateName(pAction->pMod));
 
+	if(pAction->tLastExec > now) {
+		/* if we are traveling back in time, reset tLastExec */
+		pAction->tLastExec = (time_t) 0;
+	}
 	/* now check if we need to drop the message because otherwise the action would be too
 	 * frequently called. -- rgerhards, 2008-04-08
 	 */
