From f88f5d4f7e57a6f41b01d4e06f8d40dee4c0dbad Mon Sep 17 00:00:00 2001
From: Tomas Heinrich <theinric@redhat.com>
Date: Fri, 20 Mar 2015 18:17:05 +0100
Subject: [PATCH] imjournal: fix default message priority

---
 plugins/imjournal/imjournal.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/plugins/imjournal/imjournal.c b/plugins/imjournal/imjournal.c
index 7c2c2fa..b5866e8 100644
--- a/plugins/imjournal/imjournal.c
+++ b/plugins/imjournal/imjournal.c
@@ -97,8 +97,8 @@ static struct cnfparamblk modpblk =
 	};
 
 #define DFLT_persiststateinterval 10
-#define DFLT_SEVERITY pri2fac(LOG_NOTICE)
-#define DFLT_FACILITY pri2sev(LOG_USER)
+#define DFLT_SEVERITY pri2sev(LOG_NOTICE)
+#define DFLT_FACILITY pri2fac(LOG_USER)
 
 static int bLegacyCnfModGlobalsPermitted = 1;/* are legacy module-global config parameters permitted? */
 
-- 
1.9.3

